temporalio/ext/src/worker.rs

```diff
@@ -0,0 +1,407 @@
+            .tuner(Arc::new(build_tuner(
+                options
+                    .child(id!("tuner"))?
+                    .ok_or_else(|| error!("Missing tuner"))?,
+            )?))
+            // TODO(cretz): workflow_failure_errors
+            // TODO(cretz): workflow_types_to_failure_errors
+            .build()
+            .map_err(|err| error!("Invalid worker options: {}", err))?;
+
+        let worker = temporal_sdk_core::init_worker(
+            &client.runtime_handle.core,
+            config,
+            client.core.clone().into_inner(),
+        )
+        .map_err(|err| error!("Failed creating worker: {}", err))?;
+        Ok(Worker {
+            core: RefCell::new(Some(Arc::new(worker))),
+            runtime_handle: client.runtime_handle.clone(),
+            activity,
+            _workflow,
+        })
+    }
+
+    pub fn async_poll_all(workers: RArray, queue: Value) -> Result<(), Error> {
+        // Get the first runtime handle
+        let runtime = workers
+            .entry::<typed_data::Obj<Worker>>(0)?
+            .runtime_handle
+            .clone();
+
+        // Create stream of poll calls
+        // TODO(cretz): Map for workflow pollers too
+        let worker_streams = workers
+            .into_iter()
+            .enumerate()
+            .filter_map(|(index, worker_val)| {
+                let worker_typed_data = RTypedData::from_value(worker_val).expect("Not typed data");
+                let worker_ref = worker_typed_data.get::<Worker>().expect("Not worker");
+                if worker_ref.activity {
+                    let worker = Some(
+                        worker_ref
+                            .core
+                            .borrow()
+                            .as_ref()
+                            .expect("Unable to borrow")
+                            .clone(),
+                    );
+                    Some(Box::pin(stream::unfold(worker, move |worker| async move {
+                        // We return no worker so the next streamed item closes
+                        // the stream with a None
+                        if let Some(worker) = worker {
+                            let res =
+                                temporal_sdk_core_api::Worker::poll_activity_task(&*worker).await;
+                            let shutdown_next = matches!(res, Err(PollActivityError::ShutDown));
+                            Some((
+                                (index, WorkerType::Activity, res),
+                                // No more worker if shutdown
+                                if shutdown_next { None } else { Some(worker) },
+                            ))
+                        } else {
+                            None
+                        }
+                    })))
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+        let mut worker_stream = stream::select_all(worker_streams);
```

Comment:
You can probably avoid making the boxed inner streams and this step by using https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.for_each on a stream made from https://docs.rs/futures/latest/futures/stream/fn.iter.html of worker refs and then basically inlining the below stuff into here after calling the poll.

Not a big deal though, but might read a bit nicer and be slightly less overhead

Reply:
I have the logic split into Ruby thread and Rust thread because some Ruby things are needed to build the polling streams and then I combine multiple polling streams into one polling stream. I am not sure of a good way to combine multiple streams until all done without using `select_all` which requires the stream be pinned.